zst/README.md

```diff
@@ -0,0 +1,552 @@
+is received on the stream of zng records that are to be treated as a
+zst object and everything about the zst object is flushed to storage.
+
+A zst object can be stored entirely as one seekable object (e.g., an s3 object)
+or split into separate seekable objects (e.g., unix files) that are treated
+together as a single zst entity.  While the zst format provides much flexibility
+for how data is laid out, it's up to implementations to layout data
+in intelligent ways for efficient sequential read access in spite of occasional
+seeks.
+
+## The "Column Stream" Abstraction
+
+The zst data abstraction is built around a collection of "column streams".
+
+There is one column stream for each field of zng record type definition
+that appears collectively throughout the zng input data including records
+embedded within records.
+
+For example, a field that is an array of records containing other fields would
+have a column stream for the top-level array field as well as column a stream
+for each field inside of the array values.
+
+Each column stream represents a sequence of values (inclusive of unset values)
+comprising each occurrence where that value appears with respect to its context.
+
+Records are reconstructed one by one from the column streams by picking values
+from each appropriate column stream based on the type structure of the record and
+its relationship to the various column streams.  For hierarchical records
+(i.e., records inside of records, or records inside of arrays inside of records, etc),
+the reconstruction process is recursive (as described below).
+
+## The Physical Layout
+
+Given the above logical design of zst column streams, we now describe how
+zng rows are physically laid out across zst columns as a storage
+data structure.
+
+The overall layout of a zst object is comprised of the following sections,
+in this order:
+* the data section,
+* the reassembly section,
+* and the trailer.
+
+This layout is designed so that an implementation can buffer metadata in
+memory while writing column data in a natural order to the
+data section (based on the volume statistics of each column),
+then write the metadata into the reassembly section along with the trailer
+at the end.  This allows a zng stream to be converted to a zst object
+in a single pass.
+
+> That said, the layout is
+> flexible enough that an implementation may optimize the data layout with
+> additional passes or by writing the output to multiple files then then
+> merging them together (or even leaving the zst object as separate files
+> that can be collectively read and scanned by a zst implementation).
+
+### The Data Section
+
+The data section contains raw data values organized into "segments",
+where a segment is simply a seek offset and byte length relative to the
+data section.  Each segment contains a sequence of
+[primitive-type zng values](https://github.com/brimsec/zq/blob/zst/zng/docs/spec.md#5-primitive-types),
```

Comment:
We'd discussed relaxing this so that segments could have value columns, are you still 👍 on that?

Reply:
This turned out a biit more difficult than I thought.  Devil's in the details.  You need some way to configure the writer where to stop the recursive descent then you need to capture that information in the reassembly records, which then topologically diverge from the origin zng.TypeRecord.  All do-able but not trivial so I thought we could create an issue and get to this later after we have more experience with the library and what we want to do with it.

Reply:
👍 I was still hoping it might "just work" but if not deferring makes sense